### PR TITLE
Settings

### PR DESCRIPTION
**Pull request by Izzette**
# Settings

This attempts to implement the basic functionality of relatively elaborate settings we will need to aid in restaurant selection.  It is not totally functional yet, but getting close.
## JSON dependant

This package builds a setting tree from a `JSONObject`, and after modification, create a new `JSONObject` and save the current setting state.
This `JSONObject` could be stringified and pushed to the parse server as a field of User.
## File system

This package is modeled after a file system.  It will support many kind of lookups and searches.  Objects elsewhere in the application can be given access to a "directory", and all of it's members.  It can also grant rights to access any subdirectories, and so it goes.
- `SettingTree` is the container for the root of our settings.  Originally, I was planning on making it the only interface to our settings.  However, I now feel that direct access to setting objects is more sound.  Although direct access may puts a tiny bit more strain on the implementation of accessing settings elsewhere.
- `ASetting` is the abstract class all setting objects extend.  This includes `SettingEdge`, `SettingNode`, and `SettingArray`.
- `SettingEdge` is essentially a directory, it can hold any `Class<? extends ASetting>`.  It implements `Collection<ASetting>` so it is highly modular.  However, it should likely only be used to add wings when the divert from their default values.  Perhaps it is not good to implement `Collection<ASetting>` here.
- `SettingNode` is the simple kind of file, it can have value.  Techincally. the value's type is `Object`, but in practice it can only be the types that JSONObjects can hold, and those are `boolean`, `String`, `int`, `long` and `double`.
- `SettingArray` is a special type of file, one that can be indexed like an array.  It implements `Collection` so it is highly modular.
- `SettingPath` allows you to lookup an instance of `ASetting` with a known path.

Anyways, it's 5 AM and I've been coding all night.  We should talk about this later today.

### CODE REVIEW

**Comment by Izzette** [comment]
Noteably, I do not intend for this to be merged now.  I have to make a few changes.  I'll show you and implementation too.

**Comment by sjgilbert** [comment]
Sweet!  I won't merge it until you give me the word.  Let's talk on the phone so I can hear the details about what it does.

**Comment by Izzette** [comment]
For some reason there were a duplicate copy of settings, fixed it.  I also changed SettingPath to be actually useful.

File: ParseStarterProject/res/data/default_settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "!root": {
+    "@user_name": "guest"
```
Comment (Izzette) [inline ParseStarterProject/res/data/default_settings.json:3]: How can I bind the names of settings to string resources in strings.xml?

File: ParseStarterProject/src/com/parse/starter/setting/ESettingType.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.lang.reflect.Constructor;
+import java.security.InvalidParameterException;
+
+/**
+ * Created by izzi on 7/6/15.
+ */
+enum ESettingType {
+    settingsEdge(SettingEdge.class, "!"),
+    settingsNode(SettingNode.class, "@"),
```
Comment (Izzette) [inline ParseStarterProject/src/com/parse/starter/setting/ESettingType.java]: Maybe I should avoid this particular character?

**Comment by Izzette** [comment]
Implementations may look like:

`````` java
public class needsUser {
    public static String getUser() {
        return (String) ((SettingNode)SettingTree.settings.getSetting(path)).get();
    }

    private final static SettingAbsolutePath path = new SettingAbsolutePath(R.string.user_name);
}

After next commit implementations will look like:
```java
public class needsUser {
    public static String getUser() {
        return SettingTree.settings.getSetting(path).get();
    }

    private final static SettingAbsolutePath<SettingNode<String>> path = new SettingAbsolutePath<SettingNode<String>> ();
}
``````

**Comment by Izzette** [comment]
I'm ready to merge, but there's some merge conflicts now.  I'm a little confused about how your last two commits are now part of ... hmm ...

**Pull request merged**